flocker/provision/_gce.py

```diff
@@ -0,0 +1,494 @@
+                u"initializeParams": {
+                    u"sourceImage": unicode(image),
+                    u"diskType": (
+                        u"projects/{}/zones/{}/diskTypes/pd-standard".format(
+                            project, zone)
+                    ),
+                    u"diskSizeGb": unicode(disk_size)
+                }
+            }
+        ],
+        u"networkInterfaces": [
+            {
+                u"network": (
+                    u"projects/{}/global/networks/default".format(project)
+                ),
+                u"accessConfigs": [
+                    {
+                        u"name": u"External NAT",
+                        u"type": u"ONE_TO_ONE_NAT"
+                    }
+                ]
+            }
+        ],
+        u"metadata": {
+            u"items": [
+                {
+                    u"key": u"sshKeys",
+                    u"value": u"\n".join(list(
+                        u"{}:{}".format(x.username, x.public_key)
+                        for x in login_credentials
+                    ))
+                }
+            ]
+        },
+        u"description": description,
+        u"serviceAccounts": [
+            {
+                u"email": u"default",
+                u"scopes": [
+                    # This gives the image permission to do GCE api calls (like
+                    # creating and attaching block devices) with its built-in
+                    # service account
+                    u"https://www.googleapis.com/auth/compute",
+                ]
+            },
+        ],
+        u"tags": {
+            u"items": list(
+                tag for tag in tags
+            )
+        }
+    }
+
+    if startup_script:
+        gce_slave_instance_config[u"metadata"][u"items"].append(
+            {
+                u"key": u"startup-script",
+                u"value": startup_script
+            }
+        )
+    return gce_slave_instance_config
+
+
+@implementer(INode)
+class GCENode(PClass):
+    """
+    ``INode`` implementation for GCE.
+
+    :ivar unicode address: The public IP address of the instance.
+    :ivar unicode private_address: The network internal IP address of the
+        instance.
+    :ivar bytes distribution: The OS distribution of the instance.
+    :ivar unicode project: The project the instance a member of.
+    :ivar unicode zone: The zone the instance is in.
+    :ivar unicode name: The GCE name of the instance used to identify the
+        instance.
+    :ivar compute: A Google Compute Engine Service that can be used to make
+        calls to the GCE API.
+    """
+    address = field(type=bytes)
+    private_address = field(type=bytes)
+    distribution = field(type=bytes)
+    project = field(type=unicode)
+    zone = field(type=unicode)
+    name = field(type=unicode)
+    compute = field()
+
+    def get_default_username(self):
+        return bytes(_GCE_ACCEPTANCE_USERNAME)
```

Comment:
Done. Although layers are kind of tricky here. "get_default_username" doesn't really make sense in a GCE world, where it will create any user you want it to create. This is probably the best compromise though, where the username is passed from the thing that injects the public key into the VM to the object that can answer, "who should I try to log in as?".